### PR TITLE
Partial S(alpha, beta) tables

### LINKED ISSUE #895: Nuclides with partial S(alpha, beta) tables
Some reactor materials have isotopes which appear in a mixture of different molecular structures.  For example, the carbon in TREAT fuel assemblies is usually treated as a mixture of graphite and amorphus carbon.  In these cases, we need to apply S(alpha, beta) tables to only a fraction of the atoms in that material.

I've made a [branch](https://github.com/smharper/openmc/tree/partial_sab) that I believe implements all the necessary code.  I still need to update docs/relaxNG/tests, but that might take a while as I'll be out of town for a bit.  @tjlaboss, can you test this branch and make sure it gives the expected results for your problem?  To use it, set the optional `fraction` argument when calling `material.add_s_alpha_beta`.

### PR DESCRIPTION
**Pull request by smharper**
This PR closes #895.  It allows us to add S(a,b) tables which only apply to a fraction of the atoms in a material.  It does this through the `fraction` keyword argument on `Material.add_s_alpha_beta` and the optional `fraction` attribute/element in XML.

Doing this required a change to the way we store microscopic cross sections in `NuclideMicroXS` objects.  Previously, when the code detected an S(a,b) table, it would overwrite the `micro_xs % elastic` value with the elastic + inelastic values from S(a,b).  Now the S(a,b) data is stored in `micro_xs % scatter_sab` without overwriting `micro_xs % elastic`.  Those microscopic cross sections are also multiplied by `sab_frac` and `ONE - sab_frac`, respectively, to make a sort of pseudo-nuclide that is partly S(a,b) and partly not.

Now, the macroscopic cross section is computed with `micro_xs % elastic + micro_xs % scatter_sab`, and `scatter` in physics.F90 now has separate `if` blocks to handle S(a,b) scatter and non-S(a,b) elastic scatter.

The `fraction` element needed to be added to all the test inputs_true, and I added a non-unity fraction to the S(a,b) test.  I also migrated the S(a,b) test to a `PyAPITestHarness`.

### CODE REVIEW

**Comment by paulromano** [comment]
@smharper Is it possible with these changes to use two different S(a,b) tables that apply to the same nuclide in the same material? e.g., if I wanted a material with half H-in-H2O and half H-in-ZrH, would that work?

**Comment by smharper** [comment]
I don't think so.  `calculate_nuclide_xs` is only called once per nuclide in cross_section.F90 with one particular S(a,b) table index.  I think we'd have to put a `do`-loop in there to account for multiple tables per nuclide.  I'll add a check somewhere in initialization for a nuclide appearing in multiple tables and call `fatal_error`

**Comment by smharper** [comment]
The old `elastic` isn't really an elastic cross section though, right?  It includes inelastic S(a,b).

**Comment by paulromano** [comment]
Well, thermal inelastic is not 'inelastic' in the sense the we normally talk of (leaving the nuclide in an excited state). Rather, the scattering system is left in an excited state. So the problem is that there are two different notions of elastic/inelastic.

**Comment by smharper** [comment]
Okay.  I'll rework `NuclideMicroXS`

**Comment by smharper** [comment]
Are you sure we shouldn't mix the `sab_frac` into the cross sections of `NuclideMicroXS`?  If I understand you're proposal correctly, then we'd have to make this change to physics.F90:
```F90
    # Current:
    prob = micro_xs(i_nuclide) % elastic
    ...
    prob = prob + micro_xs(i_nuclide) % scatter_sab

    # Proposed:
    prob = micro_xs(i_nuclide) % elastic - micro_xs(i_nuclide) % sab_frac &
         * micro_xs(i_nuclide) % thermal
    ...
    prob = prob + micro_xs(i_nuclide) % sab_frac * micro_xs(i_nuclide) % thermal
```

**Comment by paulromano** [comment]
Yeah, I guess you're right -- we should probably account for the fraction in `thermal` and `thermal_elastic` since it avoids having to multiply the fraction downstream in physics.

**Comment by smharper** [comment]
Okay that last commit should address point 2 and the naming conventions.  I also removed `last_index_sab` because we can just use `index_sab`.

The term "thermal scattering" may be a little overloaded because it also brings the CXS/DBRC/ARES stuff to mind.  Maybe we could switch to "bound scattering"?

File: src/cross_section.F90 [hunk 1]
```diff
@@ -84,21 +88,22 @@ subroutine calculate_xs(p)
...
 
         ! Calculate microscopic cross section for this nuclide
         if (p % E /= micro_xs(i_nuclide) % last_E &
-             .or. p % sqrtkT /= micro_xs(i_nuclide) % last_sqrtkT) then
-          call calculate_nuclide_xs(i_nuclide, i_sab, p % E, i_grid, p % sqrtkT)
-        else if (i_sab /= micro_xs(i_nuclide) % last_index_sab) then
-          call calculate_nuclide_xs(i_nuclide, i_sab, p % E, i_grid, p % sqrtkT)
+             .or. p % sqrtkT /= micro_xs(i_nuclide) % last_sqrtkT &
+             .or. i_sab /= micro_xs(i_nuclide) % index_sab &
+             .or. sab_frac /= micro_xs(i_nuclide) % sab_frac) then
```
Comment (paulromano) [inline src/cross_section.F90]: If there are no S(a,b) tables, `sab_frac` is undefined which can cause this logical to be evaluated as true. I'm seeing a pretty big performance drop (25%) for problems where no partial S(a,b) tables are used and I'm guessing this is the reason why.

Comment (smharper) [inline src/cross_section.F90]: Thanks for timing it!  I hadn't noticed the drop.  That last commit should fix it.

Comment (paulromano) [inline src/cross_section.F90]: I was actually talking about the local variable `sab_frac` in `calculate_xs` which can also be undefined.

Comment (smharper) [inline src/cross_section.F90]: Whoops.  That one is also defined now.

**Comment by paulromano** [comment]
Awesome, I'm getting equivalent performance on this branch now.

**Pull request merged**